# pandas Fundamentals
McKinney Ch 5 (basics) + Ch 7 (data cleaning). Ch 6 has no exercises

In [154]:
import numpy as np
import pandas as pd

---
## Chapter 5 — pandas Basics

### Series

In [155]:
obj = pd.Series([4, 7, -5, 3])
obj
obj.array

<NumpyExtensionArray>
[4, 7, -5, 3]
Length: 4, dtype: int64

In [156]:
obj.index

RangeIndex(start=0, stop=4, step=1)

In [157]:
obj2 = pd.Series([4, 7, -5, 3], index=['d', 'b', 'a', 'c'])
obj2

d    4
b    7
a   -5
c    3
dtype: int64

In [158]:
obj2['a']
obj2[['c', 'a', 'd']]

c    3
a   -5
d    4
dtype: int64

In [159]:
obj2[obj2 > 0] #only keep positive values
obj2 * 2

d     8
b    14
a   -10
c     6
dtype: int64

In [160]:
np.exp(obj2)

d      54.598150
b    1096.633158
a       0.006738
c      20.085537
dtype: float64

A series maps index values (left column) to actual data values
- similar to a dictionary
- dict = unformatted, array style listing

In [161]:
"b" in obj2
"e" in obj2

False

In [162]:
sdata = {"Ohio": 35000, "Texas": 71000, "Oregon": 16000, "Utah": 5000}
obj3 = pd.Series(sdata)
obj3

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

In [163]:
obj3.to_dict()

{'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}

In [164]:
states = ["California", "Ohio", "Oregon", "Texas"] # California is not in sdata, so it will be NaN
obj4 = pd.Series(sdata, index=states)
obj4

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64

In [165]:
pd.isna(obj4) #check if values are NaN
pd.notna(obj4)

California    False
Ohio           True
Oregon         True
Texas          True
dtype: bool

In [166]:
obj3 + obj4 #arithmethic between different indexes. NaN is like undefined

California         NaN
Ohio           70000.0
Oregon         32000.0
Texas         142000.0
Utah               NaN
dtype: float64

In [167]:
obj4.name = "population" #name the entire Series
obj4.index.name = "state"
obj4

state
California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
Name: population, dtype: float64

In [168]:
obj
obj.index = ['Bob', 'Steve', 'Jeff', 'Ryan'] #adding an index
obj

Bob      4
Steve    7
Jeff    -5
Ryan     3
dtype: int64

### DataFrame
dictionary of Series' with the same index

In [169]:
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003], 
        "pop": [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
frame = pd.DataFrame(data)
frame

,state,year,pop
0,Ohio,2000,1.5
1,Ohio,2001,1.7
2,Ohio,2002,3.6
3,Nevada,2001,2.4
4,Nevada,2002,2.9
5,Nevada,2003,3.2


In [170]:
frame.head() #show only first 5 rows
frame.tail(3) #show only last 3 rows

,state,year,pop
3,Nevada,2001,2.4
4,Nevada,2002,2.9
5,Nevada,2003,3.2


In [171]:
pd.DataFrame(data, columns=['year', 'state', 'pop']) #specify column order

,year,state,pop
0,2000,Ohio,1.5
1,2001,Ohio,1.7
2,2002,Ohio,3.6
3,2001,Nevada,2.4
4,2002,Nevada,2.9
5,2003,Nevada,3.2


In [172]:
frame2 = pd.DataFrame(data, columns=['year', 'state', 'pop', 'debt'])
frame2

,year,state,pop,debt
0,2000,Ohio,1.5,NaN
1,2001,Ohio,1.7,NaN
2,2002,Ohio,3.6,NaN
3,2001,Nevada,2.4,NaN
4,2002,Nevada,2.9,NaN
5,2003,Nevada,3.2,NaN


In [173]:
frame2.columns

Index(['year', 'state', 'pop', 'debt'], dtype='object')

In [174]:
frame2["state"] #returns Series of a single column
frame2.year #works the same way

0    2000
1    2001
2    2002
3    2001
4    2002
5    2003
Name: year, dtype: int64

### .loc vs .iloc
special attributes to select rows and columns from a DataFrame

In [175]:
frame2.loc[1]

year     2001
state    Ohio
pop       1.7
debt      NaN
Name: 1, dtype: object

In [176]:
frame2.iloc[2]

year     2002
state    Ohio
pop       3.6
debt      NaN
Name: 2, dtype: object

iloc = uses integer
loc = uses labels (actual index values and column names) 

### Index alignment in arithmetic

In [177]:
frame2["debt"] = 16.5
frame2

,year,state,pop,debt
0,2000,Ohio,1.5,16.5
1,2001,Ohio,1.7,16.5
2,2002,Ohio,3.6,16.5
3,2001,Nevada,2.4,16.5
4,2002,Nevada,2.9,16.5
5,2003,Nevada,3.2,16.5


In [178]:
frame2["debt"] = np.arange (6.)
frame2

,year,state,pop,debt
0,2000,Ohio,1.5,0.0
1,2001,Ohio,1.7,1.0
2,2002,Ohio,3.6,2.0
3,2001,Nevada,2.4,3.0
4,2002,Nevada,2.9,4.0
5,2003,Nevada,3.2,5.0


In [179]:
val = pd.Series([-1.2, -1.5, -1.7], index=["two", "four", "five"])
frame2["debt"] = val
frame2

,year,state,pop,debt
0,2000,Ohio,1.5,NaN
1,2001,Ohio,1.7,NaN
2,2002,Ohio,3.6,NaN
3,2001,Nevada,2.4,NaN
4,2002,Nevada,2.9,NaN
5,2003,Nevada,3.2,NaN


In [180]:
frame2["eastern"] = frame2["state"] == "Ohio"
frame2
del frame2["eastern"]
frame2.columns

Index(['year', 'state', 'pop', 'debt'], dtype='object')

In [181]:
populations = {"Ohio": {2000: 1.5, 2001: 1.7, 2002: 3.6},
               "Nevada": {2001: 2.4, 2002: 2.9}}
frame3 = pd.DataFrame(populations)
frame3

,Ohio,Nevada
2000,1.5,NaN
2001,1.7,2.4
2002,3.6,2.9


In [182]:
frame3.T

,2000,2001,2002
Ohio,1.5,1.7,3.6
Nevada,NaN,2.4,2.9


In [183]:
pd.DataFrame(populations, index=[2001, 2002, 2003])

,Ohio,Nevada
2001,1.7,2.4
2002,3.6,2.9
2003,NaN,NaN


In [184]:
pdata = {"Ohio": frame3["Ohio"][:-1], 
         "Nevada": frame3["Nevada"][:2]}
pd.DataFrame(pdata)

,Ohio,Nevada
2000,1.5,NaN
2001,1.7,2.4


*slice notation*

[start:stop]
stop number is always excluded

In [185]:
frame3.index.name = "year"
frame3.columns.name = "state"
frame3 #unlike Series, DataFrame doesn't have a name attribute

state,Ohio,Nevada
year,,
2000,1.5,NaN
2001,1.7,2.4
2002,3.6,2.9


In [186]:
frame3.to_numpy()
frame2.to_numpy()

array([[2000, 'Ohio', 1.5, nan],
       [2001, 'Ohio', 1.7, nan],
       [2002, 'Ohio', 3.6, nan],
       [2001, 'Nevada', 2.4, nan],
       [2002, 'Nevada', 2.9, nan],
       [2003, 'Nevada', 3.2, nan]], dtype=object)

In [187]:
obj = pd.Series(np.arange(3), index=['a', 'b', 'c'])
index = obj.index
index
index[1:] 

Index(['b', 'c'], dtype='object')

In [188]:
labels = pd.Index(np.arange(3))
labels

Index([0, 1, 2], dtype='int64')

In [189]:
obj2 = pd.Series([1.5, -2.5, 0], index=labels)
obj2
obj2.index is labels

True

Index are kinda like arrays

In [190]:
frame3
frame3.columns
"Ohio" in frame3.columns #is Ohio in the columns of frame3
2003 in frame3.index #is 2003 in the index of frame3 (no)

False

In [191]:
pd.Index(["foo", "foo", "bar", "bar"])

Index(['foo', 'foo', 'bar', 'bar'], dtype='object')

### Reindexing

In [192]:
obj = pd.Series([4.5, 7.2, -5.3, 3.6], index=['d', 'b', 'a', 'c'])
obj

d    4.5
b    7.2
a   -5.3
c    3.6
dtype: float64

In [193]:
obj2 = obj.reindex(['a', 'b', 'c', 'd', 'e']) #reindexing to a new set of indexes
obj2

a   -5.3
b    7.2
c    3.6
d    4.5
e    NaN
dtype: float64

In [194]:
obj3 = pd.Series(['blue', 'purple', 'yellow'], index=[0, 2, 4])
obj3

0      blue
2    purple
4    yellow
dtype: object

In [195]:
obj3.reindex(np.arange(6), method="ffill") #forward fill method for reindexing

0      blue
1      blue
2    purple
3    purple
4    yellow
5    yellow
dtype: object

In [196]:
frame = pd.DataFrame(np.arange(9).reshape((3, 3)), 
                     index=['a', 'c', 'd'], 
                     columns=['Ohio', 'Texas', 'California'])
frame

,Ohio,Texas,California
a,0,1,2
c,3,4,5
d,6,7,8


In [197]:
frame2 = frame.reindex(index=['a', 'b', 'c', 'd'])
frame2

,Ohio,Texas,California
a,0.0,1.0,2.0
b,NaN,NaN,NaN
c,3.0,4.0,5.0
d,6.0,7.0,8.0


In [198]:
states = ["Texas", "Utah", "California"]
frame.reindex(columns=states) #reindex with a keyword

,Texas,Utah,California
a,1,NaN,2
c,4,NaN,5
d,7,NaN,8


In [199]:
frame.reindex(states, axis="columns") #same way to do ^^ with axis keyword instead

,Texas,Utah,California
a,1,NaN,2
c,4,NaN,5
d,7,NaN,8


In [200]:
frame.loc[["a", "d", "c"], ["California", "Texas"]] #reindexing with loc, changing both rows and columns at once

,California,Texas
a,2,1
d,8,7
c,5,4


frame.loc only works if all the new index labels already exist
- reindex adds NaN for missing labels

### Dropping Entries from an Axis

In [201]:
obj = pd.Series(np.arange(5.), index=['a', 'b', 'c', 'd', 'e'])
obj

a    0.0
b    1.0
c    2.0
d    3.0
e    4.0
dtype: float64

In [202]:
new_obj = obj.drop("c") #drop an index (row c) from the Series
new_obj
obj.drop(["d", "c"])

a    0.0
b    1.0
e    4.0
dtype: float64

In [203]:
data = pd.DataFrame(np.arange(16).reshape((4, 4)),
                    index=['Ohio', 'Colorado', 'Utah', 'New York'],
                    columns=['one', 'two', 'three', 'four'])
data

,one,two,three,four
Ohio,0,1,2,3
Colorado,4,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


In [204]:
data.drop(columns=["two"])

,one,three,four
Ohio,0,2,3
Colorado,4,6,7
Utah,8,10,11
New York,12,14,15


In [205]:
data.drop(index=["Colorado", "Ohio"])

,one,two,three,four
Utah,8,9,10,11
New York,12,13,14,15


In [206]:
data.drop("two", axis=1) #dropping columns with axis keyword instead of columns keyword like NumPy

,one,three,four
Ohio,0,2,3
Colorado,4,6,7
Utah,8,10,11
New York,12,14,15


In [207]:
data.drop(["two", "four"], axis="columns") #drop columns two and four

,one,three
Ohio,0,2
Colorado,4,6
Utah,8,10
New York,12,14


### Indexing, Selection, and Filtering

In [208]:
obj = pd.Series(np.arange(4.), index=['a', 'b', 'c', 'd'])
obj

a    0.0
b    1.0
c    2.0
d    3.0
dtype: float64

In [209]:
obj.loc['b'] #loc looks for label first column, not integer position. 
obj.iloc [1] #iloc looks for integer position, not label.

1.0

In [210]:
obj[2:4] #slicing

c    2.0
d    3.0
dtype: float64

In [211]:
obj[["b", "a", "d"]]

b    1.0
a    0.0
d    3.0
dtype: float64

In [212]:
obj [[1, 3]]

/var/folders/9p/c9c7v_v53bd91dc97zsmsm740000gn/T/ipykernel_5671/647580651.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  obj [[1, 3]]


b    1.0
d    3.0
dtype: float64

In [213]:
obj[obj < 2] #boolean indexing

a    0.0
b    1.0
dtype: float64

In [214]:
obj.loc[["b", "a", "d"]] #preferred to label with loc

b    1.0
a    0.0
d    3.0
dtype: float64

In [215]:
obj1 = pd.Series([1, 2, 3], index=[2, 0, 1])
obj2 = pd.Series([1, 2, 3], index=["a", "b", "c"])
obj1

2    1
0    2
1    3
dtype: int64

In [216]:
obj1[[0, 1, 2]] #sorting out the order of the index, not the position of the index.

0    2
1    3
2    1
dtype: int64

In [217]:
obj2 [[0, 1, 2]] #sorting out doesn't change order for this set

/var/folders/9p/c9c7v_v53bd91dc97zsmsm740000gn/T/ipykernel_5671/2163431866.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  obj2 [[0, 1, 2]] #sorting out doesn't change order for this set


a    1
b    2
c    3
dtype: int64

In [218]:
obj1. iloc[[0, 1, 2]] #sorting out the position of the index, not the order of the index. iloc is always integer position based.
obj2.iloc[[0, 1, 2]] #using iloc instead to reorder

a    1
b    2
c    3
dtype: int64

In [219]:
obj2.loc["b":"c"] = 5
obj2

a    1
b    5
c    5
dtype: int64

In [220]:
data = pd.DataFrame(np.arange(16).reshape((4, 4)),
                    index=['Ohio', 'Colorado', 'Utah', 'New York'],
                    columns=['one', 'two', 'three', 'four'])
data

,one,two,three,four
Ohio,0,1,2,3
Colorado,4,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


In [221]:
data["two"]
data[["three", "one"]]

,three,one
Ohio,2,0
Colorado,6,4
Utah,10,8
New York,14,12


In [222]:
data[:2] #slicing rows

,one,two,three,four
Ohio,0,1,2,3
Colorado,4,5,6,7


In [223]:
data[data["three"] > 5]

,one,two,three,four
Colorado,4,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


In [224]:
data < 5 #boolean DataFrame

,one,two,three,four
Ohio,True,True,True,True
Colorado,True,False,False,False
Utah,False,False,False,False
New York,False,False,False,False


In [225]:
data[data < 5] = 0
data

,one,two,three,four
Ohio,0,0,0,0
Colorado,0,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


#### selecting on DataFrame with loc and iloc

In [226]:
data
data.loc["Colorado"]
data.loc[["Colorado", "New York"]]

,one,two,three,four
Colorado,0,5,6,7
New York,12,13,14,15


In [227]:
data.loc["Colorado", ["two", "three"]]

two      5
three    6
Name: Colorado, dtype: int64

In [228]:
data.iloc[2] #selecting the third row of the DataFrame
data.iloc[[2, 1]]
data.iloc[2, [3, 0, 1]] #selecting third row, and columns four, one, and two (in that order)
data.iloc[[1, 2], [3, 0, 1]] #selecting rows two and three, and columns four, one, and two (in that order)

,four,one,two
Colorado,7,0,5
Utah,11,8,9


In [229]:
data.loc[:"Utah", "two"]

Ohio        0
Colorado    5
Utah        9
Name: two, dtype: int64

In [230]:
data.iloc[:, :3][data.three > 5] # selecting all rows, and columns up to three, then filtering rows where column three is greater than 5

,one,two,three
Colorado,0,5,6
Utah,8,9,10
New York,12,13,14


Boolean arrays can be used only with loc

In [231]:
data.loc[data.three >=2]

,one,two,three,four
Colorado,0,5,6,7
Utah,8,9,10,11
New York,12,13,14,15


In [232]:
ser = pd.Series(np.arange(3.))
ser

0    0.0
1    1.0
2    2.0
dtype: float64

In [233]:
ser2 = pd.Series(np.arange(3.), index=['a', 'b', 'c'])
ser2[-1] #negative indexing doesn't work with label based indexing, but works with position based indexing like iloc

/var/folders/9p/c9c7v_v53bd91dc97zsmsm740000gn/T/ipykernel_5671/3700578937.py:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  ser2[-1] #negative indexing doesn't work with label based indexing, but works with position based indexing like iloc


2.0

In [234]:
ser.iloc[-1]
ser[:2]

0    0.0
1    1.0
dtype: float64

### Pitfalls with chained indexing

In [235]:
data.loc[:, "one"] = 1
data

,one,two,three,four
Ohio,1,0,0,0
Colorado,1,5,6,7
Utah,1,9,10,11
New York,1,13,14,15


In [236]:
data.iloc[2] = 5
data

,one,two,three,four
Ohio,1,0,0,0
Colorado,1,5,6,7
Utah,5,5,5,5
New York,1,13,14,15


In [237]:
data.loc[data["four"] > 5] = 3
data

,one,two,three,four
Ohio,1,0,0,0
Colorado,3,3,3,3
Utah,5,5,5,5
New York,3,3,3,3


In [238]:
data.loc[data.three == 5, "three"] = 6
data

,one,two,three,four
Ohio,1,0,0,0
Colorado,3,3,3,3
Utah,5,5,6,5
New York,3,3,3,3


### Arithmetic and Data Alignment

In [239]:
s1 = pd.Series([7.3, -2.5, 3.4, 1.5], index=["a", "c", "d", "e"])
s2 = pd.Series([-2.1, 3.6, -1.5, 4, 3.1], 
                   index=["a", "c", "e", "f", "g"])
s1
s2

a   -2.1
c    3.6
e   -1.5
f    4.0
g    3.1
dtype: float64

In [240]:
s1 + s2 

a    5.2
c    1.1
d    NaN
e    0.0
f    NaN
g    NaN
dtype: float64

In [241]:
df1 = pd.DataFrame(np.arange(9.).reshape((3, 3)), columns=list("bcd"),
                     index=["Ohio", "Texas", "Colorado"])

df2 = pd.DataFrame(np.arange(12.).reshape((4, 3)), columns=list("bde"),
                        index=["Utah", "Ohio", "Texas", "Oregon"])

df1

df2

,b,d,e
Utah,0.0,1.0,2.0
Ohio,3.0,4.0,5.0
Texas,6.0,7.0,8.0
Oregon,9.0,10.0,11.0


In [242]:
df1 = pd.DataFrame({"A": [1, 2]})
df2 = pd.DataFrame({"B": [3, 4]})
df1
df2 
df1 + df2

,A,B
0,NaN,NaN
1,NaN,NaN


##### Arithmetic methods with fill values

In [243]:
df1 = pd.DataFrame(np.arange(12.).reshape((3, 4)), 
                   columns=list("abcd"))
df2 = pd.DataFrame(np.arange(20.).reshape((4, 5)),
                     columns=list("abcde"))
df2.loc[1, "b"] = np.nan
df1
df2
df1 + df2

,a,b,c,d,e
0,0.0,2.0,4.0,6.0,NaN
1,9.0,NaN,13.0,15.0,NaN
2,18.0,20.0,22.0,24.0,NaN
3,NaN,NaN,NaN,NaN,NaN


In [244]:
df1.add(df2, fill_value=0)

,a,b,c,d,e
0,0.0,2.0,4.0,6.0,4.0
1,9.0,5.0,13.0,15.0,9.0
2,18.0,20.0,22.0,24.0,14.0
3,15.0,16.0,17.0,18.0,19.0


In [245]:
1 / df1

,a,b,c,d
0,inf,1.000000,0.500000,0.333333
1,0.250,0.200000,0.166667,0.142857
2,0.125,0.111111,0.100000,0.090909


In [246]:
df1.rdiv(1) #same as 1/df1, but with reverse division so that the fill value of 0 is replaced with 1/0 = inf instead of 0/1 = 0

,a,b,c,d
0,inf,1.000000,0.500000,0.333333
1,0.250,0.200000,0.166667,0.142857
2,0.125,0.111111,0.100000,0.090909


In [247]:
df1.reindex(columns=df2.columns, fill_value=0)

,a,b,c,d,e
0,0.0,1.0,2.0,3.0,0
1,4.0,5.0,6.0,7.0,0
2,8.0,9.0,10.0,11.0,0


In [248]:
arr = np.arange(12.).reshape((3, 4))
arr

array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]])

In [ ]:
arr[0]

array([0., 1., 2., 3.])

In [257]:
arr - arr[0] #subtracting the first row of the array from all rows, broadcasting the subtraction across the array

array([[0., 0., 0., 0.],
       [4., 4., 4., 4.],
       [8., 8., 8., 8.]])

In [258]:
frame = pd.DataFrame(np.arange(12.).reshape((4, 3)),
                        columns=list("bde"),
                        index=["Utah", "Ohio", "Texas", "Oregon"])
series = frame.iloc[0]
frame

,b,d,e
Utah,0.0,1.0,2.0
Ohio,3.0,4.0,5.0
Texas,6.0,7.0,8.0
Oregon,9.0,10.0,11.0


In [ ]:
series #since iloc was set to 0, displays the first row

b    0.0
d    1.0
e    2.0
Name: Utah, dtype: float64

In [260]:
frame - series

,b,d,e
Utah,0.0,0.0,0.0
Ohio,3.0,3.0,3.0
Texas,6.0,6.0,6.0
Oregon,9.0,9.0,9.0


In [262]:
series2 = pd.Series(np.arange(3), index=["b", "e", "f"])
series2

b    0
e    1
f    2
dtype: int64

In [263]:
frame + series2

,b,d,e,f
Utah,0.0,NaN,3.0,NaN
Ohio,3.0,NaN,6.0,NaN
Texas,6.0,NaN,9.0,NaN
Oregon,9.0,NaN,12.0,NaN


In [264]:
series3 = frame["d"]
frame

,b,d,e
Utah,0.0,1.0,2.0
Ohio,3.0,4.0,5.0
Texas,6.0,7.0,8.0
Oregon,9.0,10.0,11.0


In [265]:
series3

Utah       1.0
Ohio       4.0
Texas      7.0
Oregon    10.0
Name: d, dtype: float64

---
### Exercises — Ch 5

**Build and index a vitals DataFrame**

Construct a DataFrame with columns `['patient_id', 'hr', 'spo2', 'temp']` and 10 rows of made-up values.
Use `.loc` to select rows where `hr > 100`. Use `.iloc` to select the first 3 rows and first 2 columns.
Confirm the results are what you expected.

**Index alignment**

Make two Series:
```
a = pd.Series([1, 2, 3], index=['a','b','c'])
b = pd.Series([10, 20], index=['b','c'])
```
Add them. Predict what happens to index `'a'` before running.

---
## Chapter 7 — Data Cleaning

### Missing data — dropna / fillna / isnull

In [250]:
# type code as you read — predict output before running

### Outlier detection and clip

In [251]:
# type code as you read — predict output before running

### pd.cut / pd.qcut

In [252]:
# type code as you read — predict output before running

### apply / map

In [253]:
# type code as you read — predict output before running

### pd.get_dummies

In [254]:
# type code as you read — predict output before running

---
### Exercises — Ch 7

**Missing vitals**

Build a DataFrame with columns `['hr', 'spo2', 'temp']` and 20 rows.
Randomly set ~15% of values to NaN using `rng.choice`.
Count missing values per column. Then fill `hr`/`spo2` with forward-fill and `temp` with the column mean.
Verify no NaNs remain.

**Outlier clip**

Given `spo2 = pd.Series([98, 102, 75, 99, -3, 95, 101, 88])`,
identify values outside physiologic range (70–100) using a boolean mask.
Replace them with NaN, then forward-fill.

**Binning HR into clinical labels**

Given `hr = pd.Series([55, 78, 105, 160, 190, 95, 130])`,
use `pd.cut` with `bins=[0, 60, 100, 160, 300]` and `labels=['brady','low-normal','normal','tachy']`.

**Dummy variables**

Given a DataFrame with a `'sex'` column containing `'M'`/`'F'` and a `'ga_group'` column with `'preterm'`/`'term'`,
use `pd.get_dummies` with `drop_first=True` and confirm the output shape.